clustered/integration-test/src/test/java/org/ehcache/clustered/BasicCacheOpsMultiThreadedTest.java

```diff
@@ -0,0 +1,208 @@
+    Cache<String, Boolean> synCache = cacheManager.getCache(SYN_CACHE_NAME, String.class, Boolean.class);
+    Cache<Long, CustomValue> customValueCache = cacheManager.getCache(CLUSTERED_CACHE_NAME, Long.class, CustomValue.class);
+    String firstClientStartKey = "first_client_start", firstClientEndKey = "first_client_end";
+    if (synCache.putIfAbsent(firstClientStartKey, true) == null) {
+      silentPuts(customValueCache);
+      customValueCache.put(1L, customValue);
+      assertThat(customValueCache.get(1L), is(customValue));
+      synCache.put(firstClientEndKey, true);
+    } else {
+      //wait for the first client to finish
+      silentPuts(customValueCache);
+      int retry = 0, maxRetryCount = 30;
+      while (++retry <= maxRetryCount && synCache.get(firstClientEndKey) == null) {
+        Thread.sleep(1000L);
+      }
+
+      if (retry > maxRetryCount) {
+        Assert.fail("Couldn't find " + firstClientEndKey + " in synCache after " + maxRetryCount + " retries!");
+      }
+
+      assertThat(customValueCache.get(1L), is(customValue));
+    }
+  }
+
+  private static PersistentCacheManager createCacheManager(URI clusterURI) {
+    ServerSideConfigurationBuilder serverSideConfigBuilder = ClusteringServiceConfigurationBuilder
+        .cluster(clusterURI.resolve(CACHE_MANAGER_NAME))
+          .timeouts(TimeoutsBuilder.timeouts().read(Duration.ofSeconds(20)).write(Duration.ofSeconds(30)))
+        .autoCreate()
+        .defaultServerResource(PRIMARY_SERVER_RESOURCE_NAME);
+
+    ResourcePool resourcePool = ClusteredResourcePoolBuilder
+      .clusteredDedicated(PRIMARY_SERVER_RESOURCE_NAME, PRIMARY_SERVER_RESOURCE_SIZE, MemoryUnit.MB);
+
+    CacheManagerBuilder<PersistentCacheManager> clusteredCacheManagerBuilder = CacheManagerBuilder
+      .newCacheManagerBuilder()
+      .with(serverSideConfigBuilder)
+      .withCache(CLUSTERED_CACHE_NAME,
+        CacheConfigurationBuilder
+          .newCacheConfigurationBuilder(Long.class, CustomValue.class,
+            ResourcePoolsBuilder.newResourcePoolsBuilder()
+              .with(resourcePool))
+          .add(new ClusteredStoreConfiguration(Consistency.STRONG)))
+      .withCache(SYN_CACHE_NAME,
+        CacheConfigurationBuilder
+          .newCacheConfigurationBuilder(String.class, Boolean.class,
+            ResourcePoolsBuilder.newResourcePoolsBuilder()
+              .with(resourcePool))
+          .add(new ClusteredStoreConfiguration(Consistency.STRONG)));
+    return clusteredCacheManagerBuilder.build(false);
+  }
+
+  private void silentPuts(Cache<Long, CustomValue> customValueCache) {
+    customValueCache.put(2L, new CustomValue("value1"));
+    customValueCache.put(2L, new CustomValue("value2"));
+    customValueCache.put(3L, new CustomValue("value3"));
+    customValueCache.put(4L, new CustomValue("value4"));
+    customValueCache.put(2L, new CustomValue("value5"));
+  }
+
+  private static class CustomValue implements Serializable {
```

Comment:
Do you really need the CustomValue? Why not just putting a String? If you do, I'm pretty sure you don't need the equals and hashcode